sphinx/parsers.py

```diff
@@ -20,6 +19,32 @@
     from sphinx.application import Sphinx
 
 
+class RSTStateMachine(docutils.parsers.rst.states.RSTStateMachine):
+    def __init__(self, app: Sphinx, state_classes: Sequence[type[State]], initial_state: str,
+                 debug: bool = False):
+        self.app = app
+        super().__init__(state_classes=state_classes, initial_state=initial_state,
+                         debug=debug)
+
+    def insert_input(self, include_lines, path):
+        # First we need to combine the lines back into text so we can send it with the
+        # source-read event. In newer releases of docutils there are two lines at the end,
+        # that act as markers. We must preserve them and leave them out of the source-read
+        # event:
+        text = "\n".join(include_lines[:-2])
```

Comment:
I wish there was, but as you can see [here](https://sourceforge.net/p/docutils/code/HEAD/tree/tags/docutils-0.18/docutils/parsers/rst/directives/misc.py#l54) the text gets included, where options can dictate that only a chunk of the text is used.  See `start-line`/`end-line` and `start-after`/`end-before`.  The `rawtext` is unfortunately never assigned to a member variable so there isn't a way for us to access it. In order to make things easier in the future, then perhaps docutils maintainers could be persuaded to extend `insert_input` so that both the string lines and rawtext were passed to the function.  However, this wouldn't be great because when splitting the rawtext into lines we would now have to be careful to add the marker back in before passing it along.

My view is that without major surgery in Docutils where it add supports for an external event system, we have no other option than to become aware of this marker and handle it on the Sphinx side.  Well, there is also the option of completely re-implementing the `run` method in the `sphinx.directives.other.Include` class but that feels like a lot of maintenance.

Reply:
Yes, that would be a good idea. I actually find quite frustrating that docutils (or Sphinx actually) have some hardcoded implementations that are hard to monkey-patch without either rewriting the whole flow (which then depends on upstream changes). It'd be better if these implementations had (protected) methods that we could subclass when needed without having to bother about the rest of the implementation.

Reply:
Agreed. I do think docutils would benefit from some refactoring to make customization easier and more robust.